keys/src/multisig/mod.rs

```diff
@@ -0,0 +1,349 @@
+use std::vec;
+
+use curve25519_dalek::{
+    constants::ED25519_BASEPOINT_TABLE,
+    edwards::{CompressedEdwardsY, EdwardsPoint},
+    scalar::Scalar,
+};
+use nimiq_hash::{sha512::Sha512Hasher, Hasher};
+use nimiq_utils::key_rng::SecureGenerate;
+use rand::Rng;
+use rand_core::{CryptoRng, RngCore};
+use sha2::{self, Digest, Sha512};
+
+use self::{
+    commitment::{Commitment, CommitmentPair, Nonce},
+    error::PartialSignatureError,
+    partial_signature::PartialSignature,
+    public_key::DelinearizedPublicKey,
+};
+use crate::{KeyPair, PublicKey};
+
+pub mod address;
+pub mod commitment;
+pub mod error;
+pub mod partial_signature;
+pub mod public_key;
+
+pub const MUSIG2_PARAMETER_V: usize = 2; // Parameter used in Musig2
+
+#[cfg(feature = "serde-derive")]
+fn deserialize_scalar<'de, D>(deserializer: D) -> Result<Scalar, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let buf: [u8; commitment::Nonce::SIZE] = serde_big_array::BigArray::deserialize(deserializer)?;
+    Ok(Scalar::from_bytes_mod_order(buf))
+}
+
+#[cfg(feature = "serde-derive")]
+fn serialize_scalar<S>(value: &Scalar, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    serde_big_array::BigArray::serialize(&value.to_bytes(), serializer)
+}
+
+///
+#[derive(Clone)]
+#[cfg_attr(feature = "serde-derive", derive(serde::Serialize, serde::Deserialize))]
+pub struct CommitmentsData {
+    /// Our own nonces.
+    /// This is only needed for signing, not for calculating aggregate keys/commitments.
+    pub nonces: Option<[Nonce; MUSIG2_PARAMETER_V]>,
+    /// Our own commitments.
+    /// This is only needed for verification.
+    pub commitments: [Commitment; MUSIG2_PARAMETER_V],
```

Comment:
Shouldn't this be an `Option`, too, if it's only needed for verification? Otherwise I have to pass default Commitments in for signing anyway.

Reply:
If you use the builder, it always has the information to fill the commitments. If you know the nonces, you also know the commitments.